panel_client/panel_client.py

```diff
@@ -0,0 +1,170 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import locale
+import json
+import multiprocessing
+from Queue import Empty as EmptyQueueException
+import os
+import socket
+import struct
+import sys
+import time
+
+# Polling interval of the IO process
+MIN_LATENCY_MS = 10
+
+# characer that delimits socket messages
+DELIMITER_CHAR = '\r'
+
+
+class PanelStateBase:
+
+  def diff_states(old_state, new_state):
+    return {
+        k:v for k in new_state.iteritems()
+        if (k not in old_state) or (v != old_state[k])
+    }
+
+  def get_state_updates(self):
+    """Must implement this function for your panel.
+
+    Should return an iterable of control_id, state pairs for anything that has
+    changed since the last call.
+    """
+    raise NotImplementedError()
+
+  def get_controls(self):
+    """Must implement this function for your panel.
+
+    Should return the available controls.
+    """
+    raise NotImplementedError()
+
+  def display_message(self, message):
+    """Must implement this function for your panel.
+
+    Display a message from the server for the user."""
+
+
+def _make_update_message(update):
+  return {'message': 'set-state', 'data': {'id': update[0], 'state': update[1]}}
+
+
+def _make_announce_message(controls):
+  return {'message': 'announce', 'data': {'controls':controls}}
+
+
+def _panel_io_subprocess_main(panel_state_factory, action_queue, message_queue):
+  panel_state = panel_state_factory()
+  action_queue.put(_make_announce_message(panel_state.get_controls()))
+
+  while True:
+    for update in panel_state.get_state_updates():
```

Comment:
Is there a risk of this blocking? Not in the keyboard panel's implementation but in the actual panels, will they be polling vs. interrupt driven and if so should we issue guidance that this should only do a single poll. (Or, will the polling occur outside of this method?) Partly a question for @igor47 and @dawnho. If this does block, it could starve out messages from the server i.e. to adjust the command displayed.

Reply:
get_state_updates should be a single poll call. In the case of a panel with only "switches", it would simply consist of reading all the inputs and diffing vs the last read. 

I'm also thinking the message display should probably be in a different control loop. I think the server io loop should route 'display' messages to the display, and anything else can go on the message queue for the panel logic to interpret.